src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/InlineCompletionHandler.cs

```diff
@@ -0,0 +1,176 @@
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            if (projectionResult.LanguageKind != RazorLanguageKind.CSharp && projectionResult.LanguageKind != RazorLanguageKind.Html)
+            {
+                _logger.LogInformation($"Inline completions not supported for {projectionResult.LanguageKind}");
+                return null;
+            }
+
+            var inlineCompletionParams = new VSInternalInlineCompletionRequest
+            {
+                Context = request.Context,
+                TextDocument = new TextDocumentIdentifier { Uri = projectionResult.Uri },
+                Position = projectionResult.Position,
+            };
+
+            _logger.LogInformation($"Requesting inline completions for {projectionResult.Uri}.");
+
+            var serverKind = projectionResult.LanguageKind.ToLanguageServerKind();
+            var textBuffer = serverKind.GetTextBuffer(documentSnapshot);
+            var languageServerName = serverKind.ToLanguageServerName();
+            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<VSInternalInlineCompletionRequest, VSInternalInlineCompletionList?>(
+                textBuffer,
+                VSInternalMethods.TextDocumentInlineCompletionName,
+                languageServerName,
+                inlineCompletionParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (!ReinvocationResponseHelper.TryExtractResultOrLog(response, _logger, languageServerName, out var result))
+            {
+                return null;
+            }
+
+            _logger.LogInformation("Received result, remapping.");
+
+            if (response?.Response == null)
+            {
+                return null;
+            }
+
+            var formattingOptions = _formattingOptionsProvider.GetOptions(documentSnapshot);
+
+            var items = new List<VSInternalInlineCompletionItem>();
+            foreach (var item in response.Response.Items)
+            {
+                var containsSnippet = item.TextFormat == InsertTextFormat.Snippet;
+                var range = item.Range ?? new Range { Start = projectionResult.Position, End = projectionResult.Position };
+
+                var textEdit = new TextEdit { NewText = item.Text, Range = range };
+                var remappedEdit = await _documentMappingProvider.RemapFormattedTextEditsAsync(projectionResult.Uri, new[] { textEdit }, formattingOptions, containsSnippet, cancellationToken).ConfigureAwait(false);
```

Comment:
Do you think instead of having this overarching class as a handler in the HtmlCSharp language server we could have this as an endpoint in the Razor language server? I imagine it'd be similar to our [CodeAction endpoint](https://github.com/dotnet/razor-tooling/blob/c456c45685fe0d78039ca62e325171f851d11e50/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs#L28) that [delegates](https://github.com/dotnet/razor-tooling/blob/c456c45685fe0d78039ca62e325171f851d11e50/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs#L242-L274) back to the [client to acquire code actions](https://github.com/dotnet/razor-tooling/blob/c456c45685fe0d78039ca62e325171f851d11e50/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs#L281-L318) on the C# language server. That being said the CodeAction bits are a first-class endpoint in O# BUT we could treat this inline completion endpoint similar to the [OnAutoInsertEndpoint](https://github.com/dotnet/razor-tooling/blob/c456c45685fe0d78039ca62e325171f851d11e50/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/OnAutoInsertEndpoint.cs#L20) where we [declare all the necessary bits to interop](https://github.com/dotnet/razor-tooling/blob/c456c45685fe0d78039ca62e325171f851d11e50/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/IOnAutoInsertHandler.cs) with VS. I guess even a better example which is a custom request that delegates over to C# is @davidwengier's [breakpoint request / addition](https://github.com/dotnet/razor-tooling/pull/5987).

Some benefits of having it in the Razor language server:
- This formatting call does a StreamJsonRpc request for every response edit in order to format it. If it's all in the language server that wouldn't require any StreamJsonRpc over-the-wire hops
- We'll be able to add functionality in a single location for any future snippet expansion (I'm imagining expanding things like `@inject` in Razor
- Sets us up for adopting the new embedded language spec re-design (it'd basically be following that model exactly)
- Allows us greater control on when we want to allow C# snippet expansion. For instance, we probably don't want to allow any snippet expansion in Razor directive attributes (don't have to code this for now but it's an example of a benefit in having the endpoint in the Razor language server)